File: Amazon/Q3.cpp
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


 // } Driver Code Ends
class Solution {
  public:
    vector<int> max_of_subarrays(vector<int> arr, int n, int k) {
        // your code here
        vector<int> ans;
        std::deque<int> Q(k);
        int i;
        
        for (i=0; i<k; i++){
            while ((!Q.empty()) && arr[i] >= arr[Q.back()])
                Q.pop_back();
            Q.push_back(i);
        }
        
        for (; i<n; i++)
        {
            ans.push_back(arr[Q.front()]);
            
            while ((!Q.empty()) && Q.front() <=i - k)
                Q.pop_front();
            
            while ((!Q.empty()) && arr[i] >=arr[Q.back()])
                Q.pop_back();
 
            Q.push_back(i);
        }
        ans.push_back(arr[Q.front()]);
        return ans;
    }
};

// { Driver Code Starts.

int main() {

    int t;
    cin >> t;

    while (t--) {

        int n, k;
        cin >> n >> k;

        vector<int> arr(n);
        for (int i = 0; i < n; i++) cin >> arr[i];
        Solution ob;
        vector<int> res = ob.max_of_subarrays(arr, n, k);
        for (int i = 0; i < res.size(); i++) cout << res[i] << " ";
        cout << endl;
    }

    return 0;
}  // } Driver Code Ends

File: Amazon/Q3.py
#User function Template for python3

class Solution:
    def max_of_subarrays(self,arr,n,k):
        '''
        you can use collections module here.
        :param a: given array
        :param n: size of array
        :param k: value of k
        :return: A list of required values 
        '''
        #code here
        res = []
        d = deque()
    
        for i in range(k):
            while len(d) and arr[i]>=arr[d[-1]]:
                d.pop()
            d.append(i)
    
        for i in range(k,n):
            res.append (arr[d[0]])
            while len(d) and d[0]<=i-k:
                d.popleft()
            while len(d) and arr[i]>=arr[d[-1]]:
                d.pop()
            d.append(i)
    
        res.append (arr[d[0]])
        d.popleft()
        return res

#{ 
#  Driver Code Starts
#Initial Template for Python 3

import atexit
import io
import sys
from collections import deque

#Contributed by : Nagendra Jha

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER

@atexit.register

def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases) :
        n,k = map(int,input().strip().split())
        arr = list(map(int,input().strip().split()))
        ob=Solution()
        res = ob.max_of_subarrays(arr,n,k)
        for i in range (len (res)):
            print (res[i], end = " ")
        print()
# } Driver Code Ends
